From ae51522378891846bc6c772a3e734219cf1e04a2 Mon Sep 17 00:00:00 2001
From: Sarika Sharma <quic_sarishar@quicinc.com>
Date: Wed, 19 Jun 2024 12:47:24 +0530
Subject: [PATCH] wifi: ath12k: add WMI service bit for MEC support

Changes are done to check WMI_SERVICE_MEC_AGING_TIMER_SUPPORT service
bit set to know that multicast echo check(MEC) feature is supported
by firmware.

If host is enabling MEC feature with MEC agening timer, without
checking for firmware support, firmware crash is happening.

Hence, add WMI_SERVICE_MEC_AGING_TIMER_SUPPORT service bit check
before enabling MEC in host.

Patch-dependency: 905-wifi-ath12k-Add-support-for-MEC.patch

Signed-off-by: Sarika Sharma <quic_sarishar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/dp.c  | 6 +++++-
 drivers/net/wireless/ath/ath12k/wmi.h | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 605dd74..c921f46 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1567,6 +1567,7 @@ void ath12k_dp_vdev_tx_attach(struct ath12k *ar, struct ath12k_link_vif *arvif)
 {
 	struct ath12k_base *ab = ar->ab;
 	struct ath12k_vif *ahvif = arvif->ahvif;
+	bool mec_support;
 
 	arvif->tcl_metadata = HTT_TCL_META_DATA_GET(1, HTT_TCL_META_DATA_TYPE) |
 			      HTT_TCL_META_DATA_GET(arvif->vdev_id,
@@ -1594,7 +1595,10 @@ void ath12k_dp_vdev_tx_attach(struct ath12k *ar, struct ath12k_link_vif *arvif)
 			    FIELD_PREP(HAL_TCL_DATA_CMD_INFO4_CACHE_SET_NUM,
 			    	       arvif->ast_hash);
 
-	if (ahvif->vdev_type == WMI_VDEV_TYPE_STA) {
+	mec_support = test_bit(WMI_SERVICE_MEC_AGING_TIMER_SUPPORT, ab->wmi_ab.svc_map);
+
+	if (ahvif->vdev_type == WMI_VDEV_TYPE_STA &&
+	    ath12k_frame_mode == ATH12K_HW_TXRX_ETHERNET && mec_support) {
 		ath12k_wmi_pdev_set_timer_for_mec(ar, arvif->vdev_id,
 						  WMI_PDEV_MEC_AGING_TIMER_THRESHOLD_VALUE);
 		ath12k_hal_tx_vdev_mcast_ctrl_set(ab, arvif->vdev_id,
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index b7a1f66..a506f6f 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -2462,6 +2462,7 @@ enum wmi_tlv_service {
 	WMI_TLV_SERVICE_SW_PROG_DFS_SUPPORT = 395,
 	WMI_TLV_SERVICE_THERM_THROT_POUT_REDUCTION = 410,
 	WMI_SERVICE_WDS_NULL_FRAME_SUPPORT = 421,
+	WMI_SERVICE_MEC_AGING_TIMER_SUPPORT = 423,
 	WMI_MAX_EXT2_SERVICE,
 };
 
-- 
2.34.1

